kairos-contracts/demo-contract/contract/src/main.rs

```diff
@@ -0,0 +1,130 @@
+// The contract purse will be created in contract context so that it is "owned" by the contract
+// rather than the installing account.
+#[no_mangle]
+pub extern "C" fn init() {
+    if runtime::get_key(KAIROS_DEPOSIT_PURSE).is_some() {
+        runtime::revert(DepositError::AlreadyInitialized);
+    }
+
+    // initialize event schema
+    let schemas = Schemas::new().with::<Deposit>();
+    casper_event_standard::init(schemas);
+
+    let new_deposit_purse: URef = system::create_purse();
+    runtime::put_key(KAIROS_DEPOSIT_PURSE, new_deposit_purse.into());
+}
+
+#[no_mangle]
+pub extern "C" fn get_purse() {
+    let deposit_purse: URef = runtime::get_key(KAIROS_DEPOSIT_PURSE)
+        .unwrap_or_revert_with(DepositError::MissingKeyDepositPurse)
+        .into_uref()
+        .unwrap_or_revert();
+    let reference_to_deposit_purse_with_restricted_access =
+        deposit_purse.with_access_rights(AccessRights::ADD);
+    runtime::ret(
+        CLValue::from_t(reference_to_deposit_purse_with_restricted_access)
+            .unwrap_or_revert_with(DepositError::FailedToReturnContractPurseAsReference),
+    );
+}
+
+// Batch submission with host verifier: Unimplemented
+// This entry point will verify a batch proof and update the trie root
+// TODO: Store the trie root under a URef in this deposit contract
+// QUESTION: What is the initial root and state of the trie?
+#[no_mangle]
+pub extern "C" fn submit_batch() {
+    // receive some serialized proof
+    // the proof will include a public journal with the updated trie root
+    let _serialized_proof: Vec<u8> = runtime::get_named_arg("serialized_proof");
+
+    // TODO: VERIFY proof and UPDATE the trie root
+    // TODO: UPDATE last_processed_deposit_counter by `n`, where `n` is the amount of deposits processed in this batch.
+
+    // OPEN QUESTIONS
+    // How are L1 deposits verified when proving?
+    // Are deposits and transfers private or public inputs to the circuit / guest?
+}
+
+// Entry point called by a user through session code to deposit funds.
+// Due to Casper < 2.0 purse management and access control, it is necessary that
+// a temporary purse is funded and passed to the deposit contract, since this is
+// the only secure method of making a payment to a contract purse.
+#[no_mangle]
+pub extern "C" fn deposit() {
+    let temp_purse: URef = runtime::get_named_arg(RUNTIME_ARG_TEMP_PURSE);
+    let amount: U512 = runtime::get_named_arg(RUNTIME_ARG_AMOUNT);
+    let deposit_purse_uref: URef = runtime::get_key(KAIROS_DEPOSIT_PURSE)
+        .unwrap_or_revert_with(DepositError::MissingKeyDepositPurse)
+        .into_uref()
+        .unwrap_or_revert_with(ApiError::UnexpectedKeyVariant);
+    system::transfer_from_purse_to_purse(temp_purse, deposit_purse_uref, amount, None)
+        .unwrap_or_revert();
+
+    let new_deposit_record: Deposit = Deposit {
+        account: get_immediate_caller().unwrap_or_revert(),
+        amount,
+        timestamp: None,
```

Comment:
Are we going to use the timestamp at all?

Reply:
Good question, @Avi-D-coder, @koxu1996 what do you think, should I remove it?